src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,129 @@
+(defpackage #:coalton-impl/codegen/constant-propagation
+  (:use
+   #:cl
+   #:coalton-impl/codegen/ast)
+  (:local-nicknames
+   (#:tc #:coalton-impl/typechecker))
+  (:import-from
+   #:coalton-impl/codegen/traverse
+   #:action
+   #:count-applications
+   #:count-nodes
+   #:make-traverse-let-action-skipping-cons-bindings
+   #:*traverse*
+   #:traverse
+   #:traverse-with-binding-list)
+  (:export
+   #:propagate-constants))
+
+(in-package #:coalton-impl/codegen/constant-propagation)
+
+(defun constant-var-value (var constant-bindings &key no-error)
+  (declare (type symbol var)
+           (values (or node null) &optional))
+  (let ((may-be-value (cdr (assoc var constant-bindings))))
+    (cond (may-be-value
+           ;; Whenever it exists, it is always a NODE (thus, non-NIL)
+           may-be-value)
+          (no-error
+           nil)
+          (t
+           (error "Expected VAR ~S to be a constant but is not" var)))))
+
+(defun propagate-constants (node env)
+  (declare (optimize debug))
+  (labels ((constant-node-p (node constant-bindings)
+             ;; FIXME: Do we need more nodes classified as constants?
```

Comment:
It might be good to treat things like `(lisp IFix () cl:most-positive-fixnum)` as constants, but idk how hard it is to dig into a `node-lisp` to check

Reply:
maybe a simple cl:constantp call on the innards of node-lisp. might yield false negatives but that's fine.

Reply:
Given that `alexandria:define-constant` accepts a `:test` argument, should constants be treated as domain specific? So, we will have `*constant-predicates*` as a list of predicates. This would be set to a certain default, but users can modify it per their application.

Reply:
what do you have in mind for the contents of that list?

Reply:
When I wrote about `*constant-predicates*`, I was thinking that the contents of current `constant-node-p` can be broken into several predicates, including `node-literal-p` `constant-var-p` etc.

But, I guess we first need a better notion of a constant. Some related thoughts include: 

1. Does coalton have a predicate that checks identity of objects? 
2. What do the node-literals actually compile to?

For instance, suppose `==` checked for the identity of objects. Would the results of the `==-caller` below depend on whether or not constant propagation takes place? (You can replace the `"hello"` with something more commonly mutable, like a single-float array.)

```lisp
(coalton-toplevel  
  (define (==-caller)
    (let ((x "hello"))
      (== x x))))
```

Reply:
There's an unsafe equality predicate: `UNSAFE-POINTER-EQ?`

Coalton hasn't been more definitive about identity. The spirit of `==` is definitely that it should be a value-equality.

Reply:
About `*constant-predicates*`, this involves replacing the current `constant-node-p` with approximately following:

```lisp
(declaim (type list *constant-predicates*))
(defvar *constant-predicates* nil
  "A list of two-argument functions. Each function should take a NODE and a list of CONSTANT-BINDINGS and return
- NIL if the NODE is not constant
- the constant node corresponding to NODE if it is a constant")

(defun constant-node-p (node constant-bindings)
  (declare (type node node)
           (type list constant-bindings))
  (loop :for predicate :in *constant-predicates*
          :thereis (funcall predicate node constant-bindings)))

(defmacro define-constant-predicate (name (node-var &optional (constant-bindings-var nil cbp))
                                     &body body)
  (let ((constant-bindings-var (or constant-bindings-var (gensym "CONSTANT-BINDINGS"))))
    `(progn
       (defun ,name (,node-var ,constant-bindings-var)
         (declare (type node ,node-var))
         ,(unless cbp `(declare (ignore ,constant-bindings-var)))
         ,@body)
       (pushnew ',name *constant-predicates*))))

(define-constant-predicate node-literal-constant-predicate (node)
  (when (node-literal-p node)
    node))
(define-constant-predicate node-variable-constant-p (node constant-bindings)
  (when (node-variable-p node)
    (constant-var-value (node-variable-value node) constant-bindings :no-error t)))
(define-constant-predicate node-lisp-constant-p (node)
  (when (node-lisp-p node)
    (cl:constantp (node-lisp-form node))))
```

There would be another predicate bound inside `propagate-constants` which would check whether a given node-variable is bound to a instance-codegen-sym.

Reply:
I still don't quite understand in what cases different things would need to be considered constants, so an explicit example of that could be helpful. But that code looks generally reasonable, aside from some inconsistency like `constant-node-p` returning type `(values (or boolean node) &optional)`

Reply:
Let me try, but I think I am only able to think about it in slightly roundabout ways. 

So, there seems to be a difference between how SLIME/SBCL (sbcl-2.4.5 atleast) treats literal objects during compilation from an emacs buffer vs evaluation in the REPL. For example, if you compile the below in a buffer `(coalton (pointer-eq-caller))` returns `T`. If you evaluate the below in a REPL, `(coalton (pointer-eq-caller))` returns `NIL`.

```lisp
(coalton-toplevel
  (define (pointer-eq-caller)
    (let ((x "hello")
          (y "hello"))
      (unsafe-pointer-eq? x y))))
```

Fortunately, it seems coalton already recognizes this and the pointer equality check is therefore called `unsafe`.

When I originally wrote about the extensible `*constant-predicates*`, I also wanted to provide ways for constant propagation to be more inclusive. For instance, if there were a coalton type corresponding to `cffi:foreign-pointer`, then `(cffi:null-pointer)` should be treated as a constant for all purposes. Similarly, objects of a `color` type defined by

```lisp
(define-type color
  (make-color u8 u8 u8))
```

would also ideally be treated as a constant, since the color only depends on the RGB values. All of this probably involves looking at a `node-application` or `node-direct-application` and then determine if it satisfies the (domain/task-specific) constraints to be a constant.

On the other hand, we probably don't want the `node-application` or `node-direct-application` corresponding to `(cell:new ...)` to be treated as a constant.

---

I think

1. If literal objects are always immutable
2. If we disallow treating `node-application` or `node-direct-application` as immutable objects

In those cases, we probably do not need the extensible notion of `*constant-predicates*`. Incorporating 2 makes me a bit sad. But it can be overcome by providing a literal representation of an object. So, above, the `color` or `cffi:foreign-pointer` may have literal representations which get treated as constants. But `(cell:new ...)` shouldn't have a literal representation.